File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/java_to_py/code_output/PushBoxGame.py
class PushBoxGame:
    def __init__(self, map):
        self.map = [list(row) for row in map]
        self.playerRow = 0
        self.playerCol = 0
        self.targets = []
        self.boxes = []
        self.targetCount = 0
        self._is_game_over = False
        self._initGame()

    def _initGame(self):
        for row in range(len(self.map)):
            for col in range(len(self.map[row])):
                if self.map[row][col] == 'O':
                    self.playerRow = row
                    self.playerCol = col
                elif self.map[row][col] == 'G':
                    self.targets.append([row, col])
                    self.targetCount += 1
                elif self.map[row][col] == 'X':
                    self.boxes.append([row, col])

    def checkWin(self):
        boxOnTargetCount = 0
        for box in self.boxes:
            for target in self.targets:
                if box == target:
                    boxOnTargetCount += 1
        self._is_game_over = (boxOnTargetCount == self.targetCount)
        return self._is_game_over

    def move(self, direction):
        newPlayerRow = self.playerRow
        newPlayerCol = self.playerCol

        if direction == 'w':
            newPlayerRow -= 1
        elif direction == 's':
            newPlayerRow += 1
        elif direction == 'a':
            newPlayerCol -= 1
        elif direction == 'd':
            newPlayerCol += 1

        if self._isValidMove(newPlayerRow, newPlayerCol):
            if self._isBoxAt(newPlayerRow, newPlayerCol):
                boxPosition = self._getBoxPosition(newPlayerRow, newPlayerCol)
                newBoxRow = newPlayerRow + (newPlayerRow - self.playerRow)
                newBoxCol = newPlayerCol + (newPlayerCol - self.playerCol)

                if self._isValidMove(newBoxRow, newBoxCol) and not self._isBoxAt(newBoxRow, newBoxCol):
                    self._moveBox(boxPosition, newBoxRow, newBoxCol)
                    self.playerRow = newPlayerRow
                    self.playerCol = newPlayerCol
            else:
                self.playerRow = newPlayerRow
                self.playerCol = newPlayerCol

        return self.checkWin()

    def _isValidMove(self, row, col):
        return row >= 0 and row < len(self.map) and col >= 0 and col < len(self.map[row]) and self.map[row][col] != '#'

    def _isBoxAt(self, row, col):
        for box in self.boxes:
            if box == [row, col]:
                return True
        return False

    def _getBoxPosition(self, row, col):
        for box in self.boxes:
            if box == [row, col]:
                return box
        return None

    def _moveBox(self, boxPosition, newRow, newCol):
        for i in range(len(self.boxes)):
            if self.boxes[i] == boxPosition:
                self.boxes[i] = [newRow, newCol]
                break

    def isGameOver(self):
        return self._is_game_over

    def getPlayerCol(self):
        return self.playerCol

    def getPlayerRow(self):
        return self.playerRow

    def getTargets(self):
        return list(self.targets)

    def getTargetCount(self):
        return self.targetCount

    def getMap(self):
        return self.map

    def getBoxes(self):
        return list(self.boxes)

File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/java_to_cpp/code_output/PushBoxGame.cpp
#include <vector>
#include <array>

class PushBoxGame {
private:
    std::vector<std::vector<char>> map;
    int playerRow;
    int playerCol;
    std::vector<std::array<int,2>> targets;
    std::vector<std::array<int,2>> boxes;
    int targetCount;
    bool isGameOver;

    void initGame() {
        for (int row = 0; row < (int)map.size(); row++) {
            for (int col = 0; col < (int)map[row].size(); col++) {
                if (map[row][col] == 'O') {
                    this->playerRow = row;
                    this->playerCol = col;
                } else if (map[row][col] == 'G') {
                    this->targets.push_back({row, col});
                    this->targetCount++;
                } else if (map[row][col] == 'X') {
                    this->boxes.push_back({row, col});
                }
            }
        }
    }

    bool isValidMove(int row, int col) {
        return row >= 0 && row < (int)map.size() && col >= 0 && col < (int)map[row].size() && map[row][col] != '#';
    }

    bool isBoxAt(int row, int col) {
        for (auto& box : boxes) {
            if (box[0] == row && box[1] == col) {
                return true;
            }
        }
        return false;
    }

    std::array<int,2> getBoxPosition(int row, int col) {
        for (auto& box : boxes) {
            if (box[0] == row && box[1] == col) {
                return box;
            }
        }
        return {-1, -1};
    }

    void moveBox(std::array<int,2> boxPosition, int newRow, int newCol) {
        for (size_t i = 0; i < boxes.size(); i++) {
            if (boxes[i][0] == boxPosition[0] && boxes[i][1] == boxPosition[1]) {
                boxes[i] = {newRow, newCol};
                break;
            }
        }
    }

public:
    PushBoxGame(const std::vector<std::vector<char>>& map)
        : map(map), playerRow(0), playerCol(0), targetCount(0), isGameOver(false) {
        initGame();
    }

    bool checkWin() {
        int boxOnTargetCount = 0;
        for (auto& box : boxes) {
            for (auto& target : targets) {
                if (box[0] == target[0] && box[1] == target[1]) {
                    boxOnTargetCount++;
                }
            }
        }
        isGameOver = (boxOnTargetCount == targetCount);
        return isGameOver;
    }

    bool move(char direction) {
        int newPlayerRow = playerRow;
        int newPlayerCol = playerCol;

        if (direction == 'w') {
            newPlayerRow -= 1;
        } else if (direction == 's') {
            newPlayerRow += 1;
        } else if (direction == 'a') {
            newPlayerCol -= 1;
        } else if (direction == 'd') {
            newPlayerCol += 1;
        }

        if (isValidMove(newPlayerRow, newPlayerCol)) {
            if (isBoxAt(newPlayerRow, newPlayerCol)) {
                std::array<int,2> boxPosition = getBoxPosition(newPlayerRow, newPlayerCol);
                int newBoxRow = newPlayerRow + (newPlayerRow - playerRow);
                int newBoxCol = newPlayerCol + (newPlayerCol - playerCol);

                if (isValidMove(newBoxRow, newBoxCol) && !isBoxAt(newBoxRow, newBoxCol)) {
                    moveBox(boxPosition, newBoxRow, newBoxCol);
                    playerRow = newPlayerRow;
                    playerCol = newPlayerCol;
                }
            } else {
                playerRow = newPlayerRow;
                playerCol = newPlayerCol;
            }
        }

        return checkWin();
    }

    bool isGameOver() const {
        return isGameOver;
    }

    int getPlayerCol() const {
        return playerCol;
    }

    int getPlayerRow() const {
        return playerRow;
    }

    std::vector<std::array<int,2>> getTargets() const {
        return targets;
    }

    int getTargetCount() const {
        return targetCount;
    }

    std::vector<std::vector<char>>& getMap() {
        return map;
    }

    std::vector<std::array<int,2>> getBoxes() const {
        return boxes;
    }
};